### ISSUE TITLE
Add support for conversion into Collection types

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#3062]
Currently, jOOQ's default data type conversion can convert from and to array types. It would be useful to be able to convert also to `Collection` types, although the collection element type information might not be available due to generic type erasure.

---

Related issue: #3060

**Comment by kag0** [comment]
Just ran into situation today where this is needed. If you're open to PRs and can point me in the right direction, I can probably put in this feature.

**Comment by lukaseder** [comment]
Thank you very much for your suggestion, @kag0! Yes, we're very open to PRs. Our only restriction is that we need you to sign this transfer of rights agreement in order for us to be able to dual license your contribution:
https://github.com/jOOQ/jOOQ/blob/master/CONTRIBUTING.md

The implementation should go in `org.jooq.tools.Convert`. (unless you meant a different conversion into collections?)

**Comment by kag0** [comment]
Sounds good.   
I'll put it in `Convert`, is there anything in there for getting type of the elements for the target of the conversion?

**Comment by lukaseder** [comment]
Hmm, I think we should definitely talk about what you're trying to do first. Can you show an example of API usage that you'd like to achieve after your change?

**Comment by kag0** [comment]
So there are two levels of functionality.
#### simple case
The simple case is easy, and can but put in `Convert` with no problem. It just converts the array to a collection with no conversion of the elements.
```java
class myClass {
  int id;
  List<String> strings;
  ... getters, setters
}
```
```sql
CREATE TABLE my_table (
  id serial,
  strings varchar[]
)
```
```java
MyClass object = myTableRecord.into(MyClass.class);
```
#### advanced case
In the advanced case, it would be nice to also convert the elements, to say, enums. However, to do this we need the erased type of element stored in `enums`. 
```java
class myClass {
  int id;
  Set<MyEnum> enums;
  ... getters, setters
}
```
```sql
CREATE TABLE my_table (
  id serial,
  enums varchar[]
)
```
```java
MyClass object = myTableRecord.into(MyClass.class);
```

So in summary: with my change, I'd like to achieve the simple use case certainly. If generic type information is available somehow through jooq, then I'd like to handle the advanced case as well; but if not then I think addressing the advanced case is best left to a separate issue/feature.

**Comment by lukaseder** [comment]
OK, perfect. I see the idea here. As a general rule of thumb, we can always live with some gray area as the `into(Class)` method is not expected to be completely type safe. It does a lot of mapping by convention, which may fail some times.

Having said so:

- Yes, `Convert` is the right place to implement this
- jOOQ doesn't have any specific type information here, because `MyClass` is entirely in your control, not in jOOQ's
- But you may be able to pull off some reflection trick. If you declare a member of type `Set<MyEnum>`, then that type is not entirely erased. With JDK 6 API, you can do:
    ```java
    import java.lang.reflect.ParameterizedType;
    import java.lang.reflect.Type;
    import java.util.Set;
    
    public class Test {
        public Set<String> set = null;
    
        public static void main(String[] args) throws Exception {
            for (Type t : ((ParameterizedType) Test.class.getField("set").getGenericType()).getActualTypeArguments())
                System.out.println(t);
        }
    }
    ```
    try it, the above will yield: `class java.lang.String`

**Comment by lukaseder** [comment]
Also, when implementing this, there are some standard implementations that should be applied:

- The standard `Collection` should be `ArrayList`
- The standard `List` should be `ArrayList`
- The standard `Set` should be `LinkedHashSet` (to preserve iteration order)

**Comment by kag0** [comment]
Sounds good. Should I be forking off of master?

**Comment by lukaseder** [comment]
Yes, please. We won't merge this to older branches.
Also, don't forget: https://github.com/jOOQ/jOOQ/blob/master/CONTRIBUTING.md

**Comment by lukaseder** [comment]
Fixed through #5796
